File: update.py
import requests
import re

# The main page where the player is hosted
SOURCE_URL = "https://elahmad.org"
FILENAME = "playlist.m3u8"

def get_token_link():
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        'Referer': 'https://elahmad.org'
    }

    try:
        # 1. Get the main page content
        session = requests.Session()
        response = session.get(SOURCE_URL, headers=headers, timeout=15)
        
        # 2. Find the iframe source (where the actual player lives)
        iframe_match = re.search(r'iframe.*?src=["\'](.*?)["\']', response.text)
        if not iframe_match:
            print("Could not find the player iframe.")
            return None
        
        iframe_url = iframe_match.group(1)
        if iframe_url.startswith('//'):
            iframe_url = 'https:' + iframe_url

        # 3. Request the player page to find the .m3u8 link with the token
        player_resp = session.get(iframe_url, headers=headers, timeout=15)
        
        # This regex looks for the .m3u8 link containing wmsAuthSign
        token_link_match = re.search(r'(https://.*?\.m3u8\?.*?wmsAuthSign=[^"\'\s>]+)', player_resp.text)
        
        if token_link_match:
            return token_link_match.group(1)
        else:
            print("Token link not found in player page.")
            
    except Exception as e:
        print(f"Error: {e}")
    return None

def write_playlist(link):
    if link:
        content = f"#EXTM3U\n#EXTINF:-1, Rotana Khalijia\n{link}"
        with open(FILENAME, "w") as f:
            f.write(content)
        print(f"Updated {FILENAME} with new token.")
    else:
        print("Update failed: No link retrieved.")

if __name__ == "__main__":
    link = get_token_link()
    write_playlist(link)

File: update.js
const puppeteer = require('puppeteer');
const fs = require('fs');

(async () => {
  // These args are crucial for running on GitHub Actions
  const browser = await puppeteer.launch({ 
    headless: "new",
    args: [
      '--no-sandbox', 
      '--disable-setuid-sandbox',
      '--disable-dev-shm-usage',
      '--disable-accelerated-2d-canvas',
      '--no-first-run',
      '--no-zygote',
      '--single-process', 
      '--disable-gpu'
    ]
  });

  const page = await browser.newPage();
  // Set a real-looking user agent so the site doesn't block the browser
  await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36');

  let m3u8Link = "";

  page.on('request', request => {
    const url = request.url();
    if (url.includes('m3u8') && url.includes('wmsAuthSign')) {
      m3u8Link = url;
    }
  });

  try {
    console.log("Opening website...");
    // elahmad often checks for referers, so we go to the base site first
    await page.goto('https://elahmad.org', { waitUntil: 'networkidle2' });
    await page.goto('https://elahmad.org/tv/live/channels.php?id=1092', { waitUntil: 'networkidle2', timeout: 60000 });
    
    console.log("Waiting 30 seconds for player to load...");
    await new Promise(r => setTimeout(r, 30000));

    if (m3u8Link) {
      const content = `#EXTM3U\n#EXTINF:-1, Rotana Khalijia\n${m3u8Link}`;
      fs.writeFileSync('playlist.m3u8', content);
      console.log("SUCCESS: Link found and saved.");
    } else {
      console.log("ERROR: Could not capture the m3u8 link. The player might not have started.");
      process.exit(1); // Tell GitHub it failed so you can see the log
    }
  } catch (err) {
    console.error("CRASH: ", err);
    process.exit(1);
  } finally {
    await browser.close();
  }
})();
